project/npda/general_functions/cohort_for_date.py

```diff
@@ -0,0 +1,25 @@
+from datetime import date
+
+
+def retrieve_cohort_for_date(date_instance: date) -> int:
+    """
+    Returns the cohort for a given date.
+    """
+    audit_start_date = date(date.today().year, 4, 1)
+    if date_instance < audit_start_date:
+        # The patient was audited in the previous year
+        days_remaining = (audit_start_date - date_instance).days
+    else:
+        # The patient was audited in the current year
+        audit_end_date = date(date.today().year + 1, 3, 31)
+        days_remaining = (audit_start_date - date_instance).days
+    total_days = (audit_end_date - audit_start_date).days
+    completed_days = total_days - days_remaining
+    if (days_remaining / completed_days) < 0.25:
```

Comment:
Just sanity checking myself here. NPDA will only have 4 cohorts? Also how does the percentage ratio of how many days remaining / completed_days dictate which cohort the patient will be in? It looks like if there is less than 25% of the days left in the cohort, then they're part of cohort 1, and this increments up by 25% each time until cohort 4 gets assigned?

Reply:
this will possibly need a test. Every audit year currently (which runs from 1st April to 31st March) has only one submission, though people often upload more than one submission as they make revisions to the original submission. The NPDA team asked that we keep a record of all submissions made for a given audit year.
They also told us that HQIP are requesting that the implement quarterly reporting for a given audit year, so this is why I have introduced the idea of a cohort. The `audit_year` therefore is the year of submission (April-March), the cohort is that same year divided into quarters. There is first a check on the date of the request, to see if it is before or after April.
Any CSV uploaded is allocated an `audit_year` and `cohort_number` based on the date of upload, though only one of these can be `submission_active` (the latest).

It is possible that this function does not do what it intended. There is already a rudimentary pytest suite, and this does not require a database to test, so possibly we could write some tests for this and amend the function as necessary, in true test-driven development style...

Reply:
Gotcha - cohort is the quarter of the year submitted! A little bit of the E12 concept of cohort was leaking into my thought process but that makes sense. Perhaps it's worth a comment in the code explaining it like you have just to avoid any confusion for other devs working between e12 and npda. Happy to do this myself as I review the PR now.

The logic for the function looks good to me - but always happy to go for TDD as a sanity check! Only suggestion I have after a review is switching the numbers for cohort around - so if you have <25% days left, then you're in the 4th quarter and as such should be cohort 4, whereas at the moment for this case you would be assigned 1. Again happy to switch this around.

This will of course require another migration and the way we've been doing migration updates in this repo, looks like we're keen to keep the number of migrations low, nuking the current number in favour of having a single tidy initial migration. Is doing this just the case of deleting this current migration once changes have been made, then running makemigrations --> migrate @eatyourpeas ?

Thanks!

Reply:
Great thoughts thank you. Go ahead and change things you feel make sense. I don't feel strongly about migrations but generally while we are moving fast and breaking things I tend to start with a fresh DB and therefore like the idea of very few migrations. As things slow down then yes we should retain the changes.

Reply:
Great thanks Simon - will do. Just to double check before I implement - we're splitting up the audit year into quarters and simply assigning patients a quarter (aka cohort) dependent on which quarter their submission date falls into?

And for completed days - are we calculating the number of days since the patient's submission (aka Today - submission_date) or the number of days into the audit year the patient is (submission_date - audit_start_date)?